nemo/collections/nlp/parts/nlp_overrides.py

```diff
@@ -0,0 +1,233 @@
+
+            # With model parallelism, multiple GPUs form a large "logical GPU"
+            # this means that data parallel groups span multiple GPUs
+            # and are non-trivial
+            device_ids = self.determine_ddp_device_ids()
+            self._model = DistributedDataParallel(
+                LightningDistributedModule(self.model),
+                device_ids=device_ids,
+                output_device=device_ids[0],
+                process_group=app_state.data_parallel_group,
+                **self._ddp_kwargs,
+            )
+
+        else:
+            super().configure_ddp()
+
+    def init_model_parallel(self, global_rank: int, world_size: int) -> None:
+        """ Initializes Megatron-LM model parallel if using model parallelism.
+
+        Args:
+            global_rank (int): the global process index.
+            world_size (int): the total number of GPUs, num_nodes * num_gpus
+            is_slurm_managing_tasks (bool, optional): is the cluster managed by SLURM.
+        """
+        app_state = AppState()
+
+        # we initialize megatron-lm model parallel and data parallel groups
+        # after initializing DDP with PTL.
+        if app_state.model_parallel_size is not None:
+            if torch.distributed.is_initialized():
+                mpu.initialize_model_parallel(app_state.model_parallel_size)
+                app_state.model_parallel_group = mpu.get_model_parallel_group()
+                app_state.data_parallel_group = mpu.get_data_parallel_group()
+                app_state.model_parallel_rank = mpu.get_tensor_model_parallel_rank()
+                app_state.data_parallel_rank = mpu.get_data_parallel_rank()
+                app_state.data_parallel_size = mpu.get_data_parallel_world_size()
+                logging.info(f'mp_rank: {app_state.model_parallel_rank}')
+                logging.info(f'dp_rank: {app_state.data_parallel_rank}')
+                # TODO: get random seed from PTL
+                seed = os.environ.get("PL_GLOBAL_SEED", 1234)
+                # random seed must be set for megatron model parallel init
+                _set_random_seed(seed)
+
+    @property
+    def distributed_sampler_kwargs(self):
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            # When using model parallel, data parallel groups are non-trivial and they
+            # correspond to the logical GPUs. This means that the GPUs that form a
+            # single logical GPU all need to get the same batch of data.
+            distributed_sampler_kwargs = dict(
+                num_replicas=app_state.data_parallel_size, rank=app_state.data_parallel_rank
+            )
+            return distributed_sampler_kwargs
+
+        else:
+            return super().distributed_sampler_kwargs
+
+
+class NLPCheckpointConnector(CheckpointConnector):
+    """ Override PTL CheckpointConnector to support model parallel checkpoints from Megatron-LM.
```

Comment:
So this class overrides exp_manager checkpoint connecter right? Doesnt it lose a lot of functionality?

Reply:
No this is overriding PTL CheckpointConnector. exp_manager overrides ModelCheckpoint. No functionality should be loss, just the added functionality of model parallel checkpoints.

Please let me know if you see some functionality being lost.

Reply:
Awesome, I think it should be fine if this is not overriding exp_manager (it also overrides some of PTLs connectors like this) so lets test it out on a cluster and if it works as expected, no issues.